### Setting up and connecting csv files

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

In [ ]:
temperature = pd.read_csv('/content/drive/MyDrive/ML csv files/wrist_temperature.csv')

In [ ]:
temperature.head()

,id,study_interval,is_weekend,day_in_study,timestamp,temperature_diff_from_baseline
0,1,2022,False,3,00:00:00,0.023913
1,1,2022,False,3,00:01:00,-0.021087
2,1,2022,False,3,00:02:00,-0.101087
3,1,2022,False,3,00:03:00,-0.196087
4,1,2022,False,3,00:04:00,-0.236087


In [ ]:
temperature.shape

(6856019, 6)

In [ ]:
hrv = pd.read_csv('/content/drive/MyDrive/ML csv files/heart_rate_variability_details.csv')

In [ ]:
hrv.head()

,id,study_interval,is_weekend,day_in_study,timestamp,rmssd,coverage,low_frequency,high_frequency
0,2,2022,True,7,23:25:00,37.058,0.763,817.292,257.772
1,2,2022,True,7,23:40:00,41.079,1.005,249.851,449.752
2,2,2022,True,7,23:45:00,42.649,0.970,817.605,413.626
3,2,2022,True,7,23:50:00,41.508,0.936,878.938,257.758
4,2,2022,True,7,23:55:00,40.232,1.006,145.109,395.130


In [ ]:
hrv.shape

(436262, 9)

In [ ]:
resting_hr = pd.read_csv('/content/drive/MyDrive/ML csv files/resting_heart_rate.csv')

In [ ]:
resting_hr.head()

,id,study_interval,is_weekend,day_in_study,value,error
0,1,2022,True,1,74.785346,100.000000
1,1,2022,False,2,80.407307,29.833838
2,1,2022,False,3,84.686869,24.267298
3,1,2022,False,4,83.852219,10.344703
4,1,2022,False,5,0.000000,0.000000


In [ ]:
resting_hr.shape

(13737, 6)

In [ ]:
vo2 = pd.read_csv('/content/drive/MyDrive/ML csv files/demographic_vo2_max.csv')

In [ ]:
vo2.head()

,id,study_interval,is_weekend,day_in_study,demographic_vo2_max,demographic_vo2_max_error,filtered_demographic_vo2_max,filtered_demographic_vo2_max_error
0,1,2022,True,1,33.79370,3.0,33.79370,3.00000
1,1,2022,False,2,31.34628,3.0,32.55987,1.51239
2,1,2022,False,3,29.48324,3.0,31.50628,1.02734
3,1,2022,False,4,29.84659,3.0,31.06774,0.79267
4,1,2022,False,5,30.21759,3.0,30.88130,0.65787


In [ ]:
vo2.shape

(11482, 8)

In [ ]:
exercise = pd.read_csv('/content/drive/MyDrive/ML csv files/exercise.csv')

In [ ]:
exercise.head()

,id,study_interval,is_weekend,start_day_in_study,start_timestamp,last_modified_day_in_study,last_modified_timestamp,original_start_day_in_study,original_start_timestamp,originalduration,...,activeduration,steps,logtype,manualvaluesspecified,heartratezones,activezoneminutes,elevationgain,hasgps,shouldfetchdetails,hasactivezoneminutes
0,1,2022,False,39,21:23:14,40,18:51:51,39,21:23:14,1127000,...,1127000,1848.0,auto_detected,"{'calories': False, 'distance': False, 'steps'...","[{'name': 'Out of Range', 'min': 30, 'max': 12...","{'totalMinutes': 1, 'minutesInHeartRateZones':...",18.288,False,False,True
1,1,2022,False,39,21:23:14,40,18:51:51,39,21:23:14,1127000,...,1127000,1848.0,auto_detected,"{'calories': False, 'distance': False, 'steps'...","[{'name': 'Out of Range', 'min': 30, 'max': 12...","{'totalMinutes': 1, 'minutesInHeartRateZones':...",18.288,False,False,True
2,1,2022,False,39,21:23:14,40,18:51:51,39,21:23:14,1127000,...,1127000,1848.0,auto_detected,"{'calories': False, 'distance': False, 'steps'...","[{'name': 'Out of Range', 'min': 30, 'max': 12...","{'totalMinutes': 1, 'minutesInHeartRateZones':...",18.288,False,False,True
3,1,2022,False,39,21:23:14,40,18:51:51,39,21:23:14,1127000,...,1127000,1848.0,auto_detected,"{'calories': False, 'distance': False, 'steps'...","[{'name': 'Out of Range', 'min': 30, 'max': 12...","{'totalMinutes': 1, 'minutesInHeartRateZones':...",18.288,False,False,True
4,1,2022,False,39,21:23:14,40,18:51:51,39,21:23:14,1127000,...,1127000,1848.0,auto_detected,"{'calories': False, 'distance': False, 'steps'...","[{'name': 'Out of Range', 'min': 30, 'max': 12...","{'totalMinutes': 1, 'minutesInHeartRateZones':...",18.288,False,False,True


In [ ]:
exercise.shape

(7282, 26)

In [ ]:
phases = pd.read_csv('/content/drive/MyDrive/ML csv files/hormones_and_selfreport.csv')

In [ ]:
phases.head()

,id,study_interval,is_weekend,day_in_study,phase,lh,estrogen,pdg,flow_volume,flow_color,...,headaches,cramps,sorebreasts,fatigue,sleepissue,moodswing,stress,foodcravings,indigestion,bloating
0,1,2022,True,1,Follicular,2.9,94.2,NaN,Not at all,Not at all,...,High,Very Low/Little,Very Low/Little,High,Low,Very Low/Little,Moderate,Very Low/Little,Very Low/Little,Very Low/Little
1,1,2022,False,2,Follicular,1.2,226.3,NaN,Not at all,Not at all,...,Very High,Very Low/Little,Very Low/Little,High,Very High,Very Low/Little,Moderate,Very Low/Little,Very Low/Little,Very Low/Little
2,1,2022,False,3,Follicular,3.5,276.8,NaN,Not at all,Not at all,...,High,Very Low/Little,Very Low/Little,Very High,Very High,Very Low/Little,Low,Very Low/Little,Very Low/Little,Very Low/Little
3,1,2022,False,4,Fertility,1.8,322.1,NaN,Not at all,Not at all,...,Very Low/Little,Very Low/Little,Very Low/Little,High,Very High,Very Low/Little,Low,Very Low/Little,Very Low/Little,Very Low/Little
4,1,2022,False,5,Fertility,4.6,244.9,NaN,Not at all,Not at all,...,Very Low/Little,Very Low/Little,Very Low/Little,High,High,Very Low/Little,Low,Very Low/Little,Very Low/Little,Very Low/Little


In [ ]:
phases.shape

(5659, 22)

### Identifying keys for merging files

Since study_interval is often different for the same id/day_in_study combination, the question is if I can simply join on id, study_interval, and day_in_study. I wan to see how day_in_study updates across two data_intervals so I will pull two rows from the temperature file with different study_interval values

In [ ]:
exercise[(exercise["id"] == 9) & (exercise["study_interval"].isin([2022, 2024]))] \
    .groupby("study_interval")["start_day_in_study"].first()

,start_day_in_study
study_interval,
2022,31
2024,910


Observing the data I find that day_in_study seems to continue and not restart across intervals (in 2024 it didn't reset back to day 1 in study), so, I will be joining on all three id, day_in_study, and study_interval.


### Handling Outliers

I wanted to try linear regression as a model so I decided to deal with outliers from the beginning.

In [ ]:
for name, f, col in [('temp',      temperature, 'temperature_diff_from_baseline'),
                     ('hrv_rmssd', hrv, 'rmssd'),
                     ('hrv_hf',    hrv, 'high_frequency'),
                     ('resting_hr',    resting_hr, 'value'),
                     ('vo2',       vo2, 'filtered_demographic_vo2_max'),
                     ('exercise',       exercise, 'activeduration'),
                     ]:
    print(name)
    print(f[col].describe())
    print()

temp
count    6.856019e+06
mean    -1.736854e+00
std      2.236996e+00
min     -1.060012e+01
25%     -3.140664e+00
50%     -1.443508e+00
75%     -4.410269e-02
max      6.474183e+00
Name: temperature_diff_from_baseline, dtype: float64

hrv_rmssd
count    436262.000000
mean         54.992779
std          34.759458
min           0.000000
25%          31.022000
50%          46.411500
75%          69.219750
max         894.285000
Name: rmssd, dtype: float64

hrv_hf
count    436262.000000
mean        948.048740
std        1303.759727
min           3.370000
25%         250.970500
50%         530.423000
75%        1140.186500
max       66964.692000
Name: high_frequency, dtype: float64

resting_hr
count    13737.000000
mean        63.751213
std         22.451326
min          0.000000
25%         64.169714
50%         69.786777
75%         74.076066
max         89.345984
Name: value, dtype: float64

vo2
count    11482.000000
mean      7822.533262
std      10199.147630
min         24.548290
25%  

In [ ]:
# Temperature
temperature = temperature[temperature['temperature_diff_from_baseline'].between(-10, 10)]

# HRV - rmssd
hrv = hrv[(hrv['rmssd'] > 0) & (hrv['rmssd'] < 400)]

# HRV - hf
hrv = hrv[(hrv['high_frequency'] > 0) & (hrv['high_frequency'] < 20000)]

# VO2
vo2 = vo2[(vo2['filtered_demographic_vo2_max'] >= 10) &
              (vo2['filtered_demographic_vo2_max'] <= 90)]

# Resting_HR
resting_hr = resting_hr[(resting_hr['value'] > 30) & (resting_hr['value'] < 200)]

# Exercise
exercise = exercise[(exercise['activeduration'] > 0)]

### Aggregating data

Now, most csv files include timestamps which stop me from having one row per id, day_in_study, and study_interval. To address that, I will aggregate the data depending on their type for each csv file.

**Features and data type**

*wrist_temperature:*
I am using "temperature_diff_from_baseline" which is a level so the aggregation method will be mean, std, min, and max.

*heart_rate_variability_details:*
I am using "rmssd" and "high_frequency" which are a levels so the aggregation method will be mean, std, min, and max. I inlcuded std because if there's variation in HRV throughout the day, that can be a strong indicative of stress.

*resting_heart_rate:*
I am using "value" but the file has no timestamp column, it's a row per study_day so non-applicable.

*demographic_vo2_max:*
I am using "filtered_demographic_vo2_max" but the file has no timestamp column, it's a row per study_day so non-applicable.

*exercise:*
I am using "activeduration" which is a duration so the aggregation method will be summation.

*hormones_and_selfreport:*
I am using "phase" but the file has no timestamp column, it's a row per study_day so non-applicable.

In [ ]:
key = ['id', 'study_interval', 'day_in_study']
temperature_daily = (
    temperature.groupby(key)
    .agg(
          mean_temp=("temperature_diff_from_baseline", "mean"),
          max_temp=("temperature_diff_from_baseline", "max"),
          min_temp=("temperature_diff_from_baseline", "min"),
          std_temp=("temperature_diff_from_baseline", "std"),
    )
    .reset_index()
)

print("Daily rows:", len(temperature_daily))
print("Key dupes:", temperature_daily.duplicated(key).sum())
temperature_daily.head()


Daily rows: 5138
Key dupes: 0


,id,study_interval,day_in_study,mean_temp,max_temp,min_temp,std_temp
0,1,2022,3,-2.410090,2.508913,-9.591087,2.954392
1,1,2022,4,-2.198042,2.056928,-9.308072,2.762839
2,1,2022,6,-1.609594,1.781803,-9.048197,2.942531
3,1,2022,7,-0.877416,4.473373,-8.916627,1.389731
4,1,2022,9,-0.642556,1.769944,-8.995056,1.598276


So now what about the days where there was only one row to begin with. Now the std column will have NaN. This will be addressed in the "handling missing data" section.

But why do I even have std? The interesting result isn't "std predicts stress", it's whether it predicts stress more in the luteal phase than the follicular phase. I get at this by looking at feature importance within each phase separately, or with phase x feature interaction terms.

****std is a candidate feature. rmssd_std, for instance, measures how much a person's HRV swung across the day — a plausible stress signal that's distinct from the daily average. You don't "use" it manually; it goes into your feature matrix, and the feature-importance/SHAP step later tells you whether it actually predicted stress.****

Also, I am including max and min because for example, a day that's calm except for one sharp temperature spike has a high max but a smooth std and this can be captured and interpreted.

Here is the same code for the other files:

In [ ]:
hrv_daily = (
    hrv.groupby(key)
    .agg(
          mean_rmssd=("rmssd", "mean"),
          max_rmssd=("rmssd", "max"),
          min_rmssd=("rmssd", "min"),
          std_rmssd=("rmssd", "std"),

          mean_high_frequency=("high_frequency", "mean"),
          max_high_frequency=("high_frequency", "max"),
          min_high_frequency=("high_frequency", "min"),
          std_high_frequency=("high_frequency", "std"),
    )
    .reset_index()
)

print("Daily rows:", len(hrv_daily))
print("Key dupes:", hrv_daily.duplicated(key).sum())
hrv_daily.head()

Daily rows: 4837
Key dupes: 0


,id,study_interval,day_in_study,mean_rmssd,max_rmssd,min_rmssd,std_rmssd,mean_high_frequency,max_high_frequency,min_high_frequency,std_high_frequency
0,2,2022,7,40.505200,42.649,37.058,2.114743,354.807600,449.752,257.758,90.739286
1,2,2022,8,46.127240,73.581,16.396,11.736847,514.074221,1591.749,33.887,293.049422
2,2,2022,9,45.416714,67.859,19.175,11.206875,491.226638,1460.219,66.467,258.162583
3,2,2022,10,44.951217,69.776,21.826,10.416914,486.741424,1258.932,86.429,254.071314
4,2,2022,11,39.762764,76.482,16.807,9.123668,363.367651,1102.649,55.975,196.405897


I am renaming the start_day_in_study column under exercise.csv to day_in_study so it matches the other files. This column does have the same purpose so renaming it is not an issue

In [ ]:
exercise = exercise.rename(columns={'start_day_in_study': 'day_in_study'})

In [ ]:

exercise_daily = (
    exercise.groupby(key)
    .agg(
          mean_activeduration=("activeduration", "mean"),
          max_activeduration=("activeduration", "max"),
          min_activeduration=("activeduration", "min"),
          std_activeduration=("activeduration", "std"),
    )
    .reset_index()
)

print("Daily rows:", len(exercise_daily))
print("Key dupes:", exercise_daily.duplicated(key).sum())
exercise_daily.head()

Daily rows: 1744
Key dupes: 0


,id,study_interval,day_in_study,mean_activeduration,max_activeduration,min_activeduration,std_activeduration
0,1,2022,1,1178000.0,1178000,1178000,0.0
1,1,2022,25,1331000.0,1331000,1331000,0.0
2,1,2022,33,973000.0,973000,973000,0.0
3,1,2022,34,1075000.0,1075000,1075000,0.0
4,1,2022,38,973000.0,973000,973000,0.0


I aggregated first then will handle missing data to avoid influencing the aggregated values with data that wasn't there, i.e. I could've used imputation to handle missing data

### Merging everything onto anchor

Since files have different numbers of rows I decided to left-join on **phases** and use it as an anchor.

Before merging I wanted to check if there were any duplicate rows in the vo2, resting_hr and phases files which didn't need aggregation and therefore didn't have a duplicate check.

In [ ]:
key = ['id', 'study_interval', 'day_in_study']

for name, f in [('temperature_daily', temperature_daily),
                ('hrv_daily', hrv_daily),
                ('exercise_daily', exercise_daily),
                ('vo2', vo2),
                ('resting_hr', resting_hr),
                ('phases', phases)]:
    d = f.duplicated(key).sum()
    print(f"{name:20s} dupes on key: {d}")

temperature_daily    dupes on key: 0
hrv_daily            dupes on key: 0
exercise_daily       dupes on key: 0
vo2                  dupes on key: 1330
resting_hr           dupes on key: 7485
phases               dupes on key: 0


There are lots of duplicates in the vo2 and resting_hr files. To fix this I will do the same aggregation steps.

In [ ]:
resting_hr_daily = (
    resting_hr.groupby(key)
    .agg(
          mean_bpm=("value", "mean"),
          max_bpm=("value", "max"),
          min_bpm=("value", "min"),
          std_bpm=("value", "std"),
    )
    .reset_index()
)

print("Daily rows:", len(resting_hr_daily))
print("Key dupes:", resting_hr_daily.duplicated(key).sum())
resting_hr_daily.head()

Daily rows: 4872
Key dupes: 0


,id,study_interval,day_in_study,mean_bpm,max_bpm,min_bpm,std_bpm
0,1,2022,1,74.785346,74.785346,74.785346,NaN
1,1,2022,2,80.407307,80.407307,80.407307,NaN
2,1,2022,3,84.686869,84.686869,84.686869,NaN
3,1,2022,4,83.852219,83.852219,83.852219,NaN
4,1,2022,6,82.077053,82.077053,82.077053,NaN


In [ ]:
vo2_daily = (
    vo2.groupby(key)
    .agg(
          mean_vo2=("filtered_demographic_vo2_max", "mean"),
          max_vo2=("filtered_demographic_vo2_max", "max"),
          min_vo2=("filtered_demographic_vo2_max", "min"),
          std_vo2=("filtered_demographic_vo2_max", "std"),
    )
    .reset_index()
)

print("Daily rows:", len(vo2_daily))
print("Key dupes:", vo2_daily.duplicated(key).sum())
vo2_daily.head()

Daily rows: 5374
Key dupes: 0


,id,study_interval,day_in_study,mean_vo2,max_vo2,min_vo2,std_vo2
0,1,2022,1,33.79370,33.79370,33.79370,NaN
1,1,2022,2,32.55987,32.55987,32.55987,NaN
2,1,2022,3,31.50628,31.50628,31.50628,NaN
3,1,2022,4,31.06774,31.06774,31.06774,NaN
4,1,2022,5,30.88130,30.88130,30.88130,NaN


In [ ]:
key = ['id', 'study_interval', 'day_in_study']

for name, f in [('temperature_daily', temperature_daily),
                ('hrv_daily', hrv_daily),
                ('exercise_daily', exercise_daily),
                ('vo2_daily', vo2_daily),
                ('resting_hr_daily', resting_hr_daily),
                ('phases', phases)]:
    d = f.duplicated(key).sum()
    print(f"{name:20s} dupes on key: {d}")

temperature_daily    dupes on key: 0
hrv_daily            dupes on key: 0
exercise_daily       dupes on key: 0
vo2_daily            dupes on key: 0
resting_hr_daily     dupes on key: 0
phases               dupes on key: 0


I want the phases variable to only be the file with the column phases, excluding the other columns.

In [ ]:
key = ['id', 'study_interval', 'day_in_study']

phases = phases[key + ['phase']]

In [ ]:
anchor = phases
anchor_rows = len(phases)

merged = (phases
          .merge(temperature_daily, on=key, how='left', validate='one_to_one')
          .merge(hrv_daily,         on=key, how='left', validate='one_to_one')
          .merge(exercise_daily,    on=key, how='left', validate='one_to_one')
          .merge(vo2_daily,    on=key, how='left', validate='one_to_one')
          .merge(resting_hr_daily,    on=key, how='left', validate='one_to_one')
          )


Checking that the anchoring didn't change the number of rows that phases had. If they're equal, we're good!

In [ ]:
print("Anchor rows:", anchor_rows)
print("Merged rows:", len(merged))
assert len(merged) == anchor_rows, "Row count changed — a join multiplied rows!"

Anchor rows: 5659
Merged rows: 5659


In [ ]:
merged

,id,study_interval,day_in_study,phase,mean_temp,max_temp,min_temp,std_temp,mean_rmssd,max_rmssd,...,min_activeduration,std_activeduration,mean_vo2,max_vo2,min_vo2,std_vo2,mean_bpm,max_bpm,min_bpm,std_bpm
0,1,2022,1,Follicular,NaN,NaN,NaN,NaN,NaN,NaN,...,1178000.0,0.000000e+00,33.79370,33.79370,33.79370,NaN,74.785346,74.785346,74.785346,NaN
1,1,2022,2,Follicular,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,32.55987,32.55987,32.55987,NaN,80.407307,80.407307,80.407307,NaN
2,1,2022,3,Follicular,-2.410090,2.508913,-9.591087,2.954392,NaN,NaN,...,NaN,NaN,31.50628,31.50628,31.50628,NaN,84.686869,84.686869,84.686869,NaN
3,1,2022,4,Fertility,-2.198042,2.056928,-9.308072,2.762839,NaN,NaN,...,NaN,NaN,31.06774,31.06774,31.06774,NaN,83.852219,83.852219,83.852219,NaN
4,1,2022,5,Fertility,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,30.88130,30.88130,30.88130,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5654,50,2024,947,Luteal,-2.638995,1.262798,-5.697202,1.929520,45.267622,64.006,...,1075000.0,1.142702e+06,26.04303,26.04303,26.04303,NaN,72.000306,81.385284,63.544417,7.455417
5655,50,2024,948,Luteal,-3.005161,0.939797,-8.875203,2.313486,45.065402,72.946,...,922000.0,1.776959e+06,26.05882,26.05882,26.05882,NaN,70.765693,81.137576,61.815119,7.236839
5656,50,2024,949,Luteal,-1.696406,1.334797,-7.285203,1.203487,42.634456,71.117,...,1065000.0,6.257862e+05,26.06572,26.06572,26.06572,NaN,69.743035,80.042849,61.165657,7.993399
5657,50,2024,950,Menstrual,-2.963090,1.234797,-7.960203,1.734385,51.653774,82.791,...,1230000.0,5.421402e+05,26.03662,26.03662,26.03662,NaN,65.783344,71.715221,58.715265,6.574034


### Handling Missing Data

Since we ensured data fell within a specific range and aggregated the data, there are no missing values within the _daily files. We can see that only std has NaN values and that's because if std is undefined for a single data point, it measures spread across multiple readings, so a day with only one reading has no spread to compute. These NaNs therefore indicate days where only one measurement was recorded, not missing data in the usual sense.

To deal with that, and because we want to maintain the std column, we fill the NaN values with 0. Filling with 0 reflects no observed spread while preserving the day's other valid statistics (mean, min, max) rather than dropping the row and losing data.

For the one row with NaN in phases, I removed it.

See below

In [ ]:
print(temperature_daily.isna().sum())
print(hrv_daily.isna().sum())
print(resting_hr_daily.isna().sum())
print(vo2_daily.isna().sum())
print(exercise_daily.isna().sum())
print(phases.isna().sum())

id                0
study_interval    0
day_in_study      0
mean_temp         0
max_temp          0
min_temp          0
std_temp          0
dtype: int64
id                     0
study_interval         0
day_in_study           0
mean_rmssd             0
max_rmssd              0
min_rmssd              0
std_rmssd              4
mean_high_frequency    0
max_high_frequency     0
min_high_frequency     0
std_high_frequency     4
dtype: int64
id                   0
study_interval       0
day_in_study         0
mean_bpm             0
max_bpm              0
min_bpm              0
std_bpm           2928
dtype: int64
id                   0
study_interval       0
day_in_study         0
mean_vo2             0
max_vo2              0
min_vo2              0
std_vo2           4287
dtype: int64
id                       0
study_interval           0
day_in_study             0
mean_activeduration      0
max_activeduration       0
min_activeduration       0
std_activeduration     460
dtype: int64
id       

In [ ]:
# seeing the row with a missing phase
print(phases[phases['phase'].isna()])

     id  study_interval  day_in_study phase
275   4            2022             6   NaN


In [ ]:
# removing the row with missing phase then remerging
phases = phases.dropna(subset=['phase'])

merged = (phases
          .merge(temperature_daily, on=key, how='left', validate='one_to_one')
          .merge(hrv_daily,         on=key, how='left', validate='one_to_one')
          .merge(exercise_daily,    on=key, how='left', validate='one_to_one')
          .merge(vo2_daily,    on=key, how='left', validate='one_to_one')
          .merge(resting_hr_daily,    on=key, how='left', validate='one_to_one')
          )

In [ ]:
# find columns that start with std
std_cols = [c for c in merged.columns if c.startswith('std')]

In [ ]:
merged

,id,study_interval,day_in_study,phase,mean_temp,max_temp,min_temp,std_temp,mean_rmssd,max_rmssd,...,min_activeduration,std_activeduration,mean_vo2,max_vo2,min_vo2,std_vo2,mean_bpm,max_bpm,min_bpm,std_bpm
0,1,2022,1,Follicular,NaN,NaN,NaN,NaN,NaN,NaN,...,1178000.0,0.000000e+00,33.79370,33.79370,33.79370,NaN,74.785346,74.785346,74.785346,NaN
1,1,2022,2,Follicular,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,32.55987,32.55987,32.55987,NaN,80.407307,80.407307,80.407307,NaN
2,1,2022,3,Follicular,-2.410090,2.508913,-9.591087,2.954392,NaN,NaN,...,NaN,NaN,31.50628,31.50628,31.50628,NaN,84.686869,84.686869,84.686869,NaN
3,1,2022,4,Fertility,-2.198042,2.056928,-9.308072,2.762839,NaN,NaN,...,NaN,NaN,31.06774,31.06774,31.06774,NaN,83.852219,83.852219,83.852219,NaN
4,1,2022,5,Fertility,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,30.88130,30.88130,30.88130,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5653,50,2024,947,Luteal,-2.638995,1.262798,-5.697202,1.929520,45.267622,64.006,...,1075000.0,1.142702e+06,26.04303,26.04303,26.04303,NaN,72.000306,81.385284,63.544417,7.455417
5654,50,2024,948,Luteal,-3.005161,0.939797,-8.875203,2.313486,45.065402,72.946,...,922000.0,1.776959e+06,26.05882,26.05882,26.05882,NaN,70.765693,81.137576,61.815119,7.236839
5655,50,2024,949,Luteal,-1.696406,1.334797,-7.285203,1.203487,42.634456,71.117,...,1065000.0,6.257862e+05,26.06572,26.06572,26.06572,NaN,69.743035,80.042849,61.165657,7.993399
5656,50,2024,950,Menstrual,-2.963090,1.234797,-7.960203,1.734385,51.653774,82.791,...,1230000.0,5.421402e+05,26.03662,26.03662,26.03662,NaN,65.783344,71.715221,58.715265,6.574034


In [ ]:
row = merged[(merged['id'] == 4) &
             (merged['study_interval'] == 2022) &
             (merged['day_in_study'] == 6)]
print(row)

Empty DataFrame
Columns: [id, study_interval, day_in_study, phase, mean_temp, max_temp, min_temp, std_temp, mean_rmssd, max_rmssd, min_rmssd, std_rmssd, mean_high_frequency, max_high_frequency, min_high_frequency, std_high_frequency, mean_activeduration, max_activeduration, min_activeduration, std_activeduration, mean_vo2, max_vo2, min_vo2, std_vo2, mean_bpm, max_bpm, min_bpm, std_bpm]
Index: []

[0 rows x 28 columns]


In [ ]:
print("Filling NaN in:", std_cols)

# filling with 0
merged[std_cols] = merged[std_cols].fillna(0)

print("Remaining NaN in std cols:", merged[std_cols].isna().sum().sum())   # should be 0

Filling NaN in: ['std_temp', 'std_rmssd', 'std_high_frequency', 'std_activeduration', 'std_vo2', 'std_bpm']
Remaining NaN in std cols: 0


In [ ]:
merged

,id,study_interval,day_in_study,phase,mean_temp,max_temp,min_temp,std_temp,mean_rmssd,max_rmssd,...,min_activeduration,std_activeduration,mean_vo2,max_vo2,min_vo2,std_vo2,mean_bpm,max_bpm,min_bpm,std_bpm
0,1,2022,1,Follicular,NaN,NaN,NaN,0.000000,NaN,NaN,...,1178000.0,0.000000e+00,33.79370,33.79370,33.79370,0.0,74.785346,74.785346,74.785346,0.000000
1,1,2022,2,Follicular,NaN,NaN,NaN,0.000000,NaN,NaN,...,NaN,0.000000e+00,32.55987,32.55987,32.55987,0.0,80.407307,80.407307,80.407307,0.000000
2,1,2022,3,Follicular,-2.410090,2.508913,-9.591087,2.954392,NaN,NaN,...,NaN,0.000000e+00,31.50628,31.50628,31.50628,0.0,84.686869,84.686869,84.686869,0.000000
3,1,2022,4,Fertility,-2.198042,2.056928,-9.308072,2.762839,NaN,NaN,...,NaN,0.000000e+00,31.06774,31.06774,31.06774,0.0,83.852219,83.852219,83.852219,0.000000
4,1,2022,5,Fertility,NaN,NaN,NaN,0.000000,NaN,NaN,...,NaN,0.000000e+00,30.88130,30.88130,30.88130,0.0,NaN,NaN,NaN,0.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5653,50,2024,947,Luteal,-2.638995,1.262798,-5.697202,1.929520,45.267622,64.006,...,1075000.0,1.142702e+06,26.04303,26.04303,26.04303,0.0,72.000306,81.385284,63.544417,7.455417
5654,50,2024,948,Luteal,-3.005161,0.939797,-8.875203,2.313486,45.065402,72.946,...,922000.0,1.776959e+06,26.05882,26.05882,26.05882,0.0,70.765693,81.137576,61.815119,7.236839
5655,50,2024,949,Luteal,-1.696406,1.334797,-7.285203,1.203487,42.634456,71.117,...,1065000.0,6.257862e+05,26.06572,26.06572,26.06572,0.0,69.743035,80.042849,61.165657,7.993399
5656,50,2024,950,Menstrual,-2.963090,1.234797,-7.960203,1.734385,51.653774,82.791,...,1230000.0,5.421402e+05,26.03662,26.03662,26.03662,0.0,65.783344,71.715221,58.715265,6.574034


As we can see there are lots of NaN values and that is a result of the left join. To fix these NaN values I will look at the percentage of missing values and based on that, decide what method is most appropriate.

In [ ]:
print((merged.isna().mean()*100).round(1)[lambda s: s > 0].sort_values(ascending=False))

max_activeduration     69.2
mean_activeduration    69.2
min_activeduration     69.2
mean_high_frequency    14.5
min_high_frequency     14.5
mean_rmssd             14.5
max_high_frequency     14.5
max_rmssd              14.5
min_rmssd              14.5
min_bpm                13.9
max_bpm                13.9
mean_bpm               13.9
max_temp                9.2
mean_temp               9.2
min_temp                9.2
mean_vo2                5.0
max_vo2                 5.0
min_vo2                 5.0
dtype: float64


Since the active_duration columns all have an extremely high missing percentage, I will drop them.

In [ ]:
# dropping the active_duration columns
merged = merged.drop(columns=['mean_activeduration', 'max_activeduration', 'min_activeduration', 'std_activeduration'])

For the rest I will use imputation but I need to first decide whether I will use statistical methods or time-series depending on the gap of missing values. **By gap I mean the typical longest-gap across participant-waves**.

Let's check temperature first:

In [ ]:
# this code finds the longest run of consecutive missing days per participant
def max_gap(s):
    return s.isna().astype(int).groupby(s.notna().cumsum()).sum().max()

print(merged.groupby(['id','study_interval'])['mean_temp'].apply(max_gap).describe())

count    62.000000
mean      3.645161
std       5.498425
min       0.000000
25%       1.000000
50%       2.000000
75%       3.000000
max      32.000000
Name: mean_temp, dtype: float64


For temperature, most gaps are short approx 2 days, but the tail is 32 which is very long meaning some participants have a very long stretch of not having any recordings. For the short gaps, I will use time-series imputation.

I will use interpolation instead of forward-fill because temperature is a signal that changes gradually and continiously and I can capture that through interpolation.

In [ ]:
key_sort = ['id', 'study_interval', 'day_in_study']
merged = merged.sort_values(key_sort)

merged['mean_temp'] = (merged
    .groupby(['id', 'study_interval'])['mean_temp']
    .transform(lambda s: s.interpolate(method='linear', limit=3, limit_area='inside')))

# filling the rest with the median

merged['mean_temp'] = merged['mean_temp'].fillna(merged['mean_temp'].median())

I will run the same process for the rest of the columns in my data and write out which method I used and why after the code.

In [ ]:
cols_to_check = ['max_temp', 'min_temp', 'mean_rmssd', 'min_rmssd', 'max_rmssd', 'mean_high_frequency', 'min_high_frequency', 'max_high_frequency',
                 'mean_bpm', 'min_bpm', 'max_bpm', 'mean_vo2', 'min_vo2', 'max_vo2']

for col in cols_to_check:
    gaps = merged.groupby(['id', 'study_interval'])[col].apply(max_gap)
    print(f"{col}: mean gap = {gaps.mean():.1f}")

max_temp: mean gap = 3.6
min_temp: mean gap = 3.6
mean_rmssd: mean gap = 10.9
min_rmssd: mean gap = 10.9
max_rmssd: mean gap = 10.9
mean_high_frequency: mean gap = 10.9
min_high_frequency: mean gap = 10.9
max_high_frequency: mean gap = 10.9
mean_bpm: mean gap = 11.6
min_bpm: mean gap = 11.6
max_bpm: mean gap = 11.6
mean_vo2: mean gap = 4.2
min_vo2: mean gap = 4.2
max_vo2: mean gap = 4.2


In [ ]:
key_sort = ['id', 'study_interval', 'day_in_study']
merged = merged.sort_values(key_sort)

merged['min_temp'] = (merged
    .groupby(['id', 'study_interval'])['min_temp']
    .transform(lambda s: s.interpolate(method='linear', limit=3, limit_area='inside')))

# filling the rest with the median

merged['min_temp'] = merged['min_temp'].fillna(merged['min_temp'].median())

In [ ]:
key_sort = ['id', 'study_interval', 'day_in_study']
merged = merged.sort_values(key_sort)

merged['max_temp'] = (merged
    .groupby(['id', 'study_interval'])['max_temp']
    .transform(lambda s: s.interpolate(method='linear', limit=3, limit_area='inside')))

# filling the rest with the median

merged['max_temp'] = merged['max_temp'].fillna(merged['max_temp'].median())

In [ ]:
median_cols = [
    'mean_vo2', 'max_vo2', 'min_vo2',
    'mean_rmssd', 'max_rmssd', 'min_rmssd',
    'mean_high_frequency', 'max_high_frequency', 'min_high_frequency',
    'mean_bpm', 'max_bpm', 'min_bpm'
]

for col in median_cols:
    merged[col] = merged[col].fillna(merged[col].median())

In [ ]:
# confirming these columns have no NaNs
print(merged[median_cols].isna().sum())

mean_vo2               0
max_vo2                0
min_vo2                0
mean_rmssd             0
max_rmssd              0
min_rmssd              0
mean_high_frequency    0
max_high_frequency     0
min_high_frequency     0
mean_bpm               0
max_bpm                0
min_bpm                0
dtype: int64


**Summary of other columns and how I handled missing data**

All the other columns had missing data but not enough to drop so I used imputation with them to depending on their gap.

For min_temp and max_temp I used the same process I did with mean_temp because the gap was still about 3.6 which is pretty small but there were still larger gaps.

For Vo2, although the gap was also small, about 4.2, Vo2 itself is a slow measurement that barely changes so interpolation won't do much, that's why I used median. I could've also used forward-fill (which keeps measurement constant) but it would've been the same as using median.

For the rest, I used median as the gaps were pretty large (e.g. 11.6 days)

In [ ]:
merged

,id,study_interval,day_in_study,phase,mean_temp,max_temp,min_temp,std_temp,mean_rmssd,max_rmssd,...,min_high_frequency,std_high_frequency,mean_vo2,max_vo2,min_vo2,std_vo2,mean_bpm,max_bpm,min_bpm,std_bpm
0,1,2022,1,Follicular,-1.652598,1.895196,-7.701870,0.000000,48.590630,82.0945,...,110.6305,0.000000,33.79370,33.79370,33.79370,0.0,74.785346,74.785346,74.785346,0.000000
1,1,2022,2,Follicular,-1.652598,1.895196,-7.701870,0.000000,48.590630,82.0945,...,110.6305,0.000000,32.55987,32.55987,32.55987,0.0,80.407307,80.407307,80.407307,0.000000
2,1,2022,3,Follicular,-2.410090,2.508913,-9.591087,2.954392,48.590630,82.0945,...,110.6305,0.000000,31.50628,31.50628,31.50628,0.0,84.686869,84.686869,84.686869,0.000000
3,1,2022,4,Fertility,-2.198042,2.056928,-9.308072,2.762839,48.590630,82.0945,...,110.6305,0.000000,31.06774,31.06774,31.06774,0.0,83.852219,83.852219,83.852219,0.000000
4,1,2022,5,Fertility,-1.903818,1.919366,-9.178134,0.000000,48.590630,82.0945,...,110.6305,0.000000,30.88130,30.88130,30.88130,0.0,70.739553,74.195901,65.672362,0.000000
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
5653,50,2024,947,Luteal,-2.638995,1.262798,-5.697202,1.929520,45.267622,64.0060,...,45.8980,209.322457,26.04303,26.04303,26.04303,0.0,72.000306,81.385284,63.544417,7.455417
5654,50,2024,948,Luteal,-3.005161,0.939797,-8.875203,2.313486,45.065402,72.9460,...,56.4780,383.967900,26.05882,26.05882,26.05882,0.0,70.765693,81.137576,61.815119,7.236839
5655,50,2024,949,Luteal,-1.696406,1.334797,-7.285203,1.203487,42.634456,71.1170,...,59.8740,237.203865,26.06572,26.06572,26.06572,0.0,69.743035,80.042849,61.165657,7.993399
5656,50,2024,950,Menstrual,-2.963090,1.234797,-7.960203,1.734385,51.653774,82.7910,...,65.4980,320.066735,26.03662,26.03662,26.03662,0.0,65.783344,71.715221,58.715265,6.574034
